# DATA

Study of a dataset cotaining various information about 32135 games published on Steam gaming plataform.

| Columna | Descripción | Ejemplo |
| -- | - | - |
| publisher | Empresa publicadora del contenido | [Ubisoft,Dovetail Games - Trains,Degica] |
| genres | Genero del contenido | [Action, Adventure, Racing, Simulation, Strategy] |
| app_name | Nombre del contenido | [Warzone, Soundtrack, Puzzle Blocks] |
| title | Titulo del contenido | [The Dream Machine: Chapter 4 , Fate/EXTELLA - Sweet Room Dream, Fate/EXTELLA - Charming Bunny] |
| url | URL de publicación del contenido | http://store.steampowered.com/app/761140/Lost_Summoner_Kitty/ |
| release_date | Fecha de lanzamiento | [2018-01-04] |
| tags | etiquetas de contenido | [Simulation, Indie, Action, Adventure, Funny, Open World, First-Person, Sandbox, Free to Play] |
| discount_price | precio de descuento | [22.66, 0.49, 0.69] |
| reviews_url | Reviews de contenido | http://steamcommunity.com/app/681550/reviews/?browsefilter=mostrecent&p=1 |
| specs | Especificaciones | [Multi-player, Co-op, Cross-Platform Multiplayer, Downloadable Content] |
| price | Precio del contenido | [4.99, 9.99, Free to Use, Free to Play] |
| early_access | acceso temprano | [False, True] |
| id | identificador unico de contenido | [761140, 643980, 670290] |
| developer | Desarrollador | [Kotoshiro, Secret Level SRL, Poolians.com] |
| sentiment | Análisis de sentimiento | [Mixed, Very Positive, Positive, 3 user reviews] |
| metascore | Score por metacritic | [80, 74, 77, 75] |



# PREPROCESSING

Prepare datasets to process 6 endpoint efficiently.

    def genero( Año: str ): Se ingresa un año y devuelve una lista con los 5 géneros más ofrecidos en el orden correspondiente.

    def juegos( Año: str ): Se ingresa un año y devuelve una lista con los juegos lanzados en el año.

    def specs( Año: str ): Se ingresa un año y devuelve una lista con los 5 specs que más se repiten en el mismo en el orden correspondiente.

    def earlyacces( Año: str ): Cantidad de juegos lanzados en un año con early access.

    def sentiment( Año: str ): Según el año de lanzamiento, se devuelve una lista con la cantidad de registros que se encuentren categorizados con un análisis de sentimiento.

                        Ejemplo de retorno: {Mixed = 182, Very Positive = 120, Positive = 278}

    def metascore( Año: str ): Top 5 juegos según año con mayor metascore.


Import relevant libraries.

In [31]:
import pandas as pd
import numpy as np
import ast
import json

## Read Data

I tried to open the json file with pd.read_json but i could not do it because of it's format.

In [32]:
# pd.read_json()

That's why I decided to use "open" and then transform each row into a dictionary. With the list of all the rows I can create a DataFrame.

In [33]:
# Open file.
file = open('datasets/steam_games.json')

allNewLines = []

# Go through each line 
for line in file:
    # Create a dictionary of each line.
    myNewdict = ast.literal_eval(line)
    # Append to a list of all the lines in the file.
    allNewLines.append(myNewdict)

rawGamesDf = pd.DataFrame(allNewLines)

In [34]:
rawGamesDf.head(2)

publisher                                         genres   
0         Kotoshiro  [Action, Casual, Indie, Simulation, Strategy]  \
1  Making Fun, Inc.           [Free to Play, Indie, RPG, Strategy]   

              app_name                title   
0  Lost Summoner Kitty  Lost Summoner Kitty  \
1            Ironbound            Ironbound   

                                                 url release_date   
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04  \
1  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   

                                                tags  discount_price   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49  \
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   

                                         reviews_url   
0  http://steamcommunity.com/app/761140/reviews/?...  \
1  http://steamcommunity.com/app/643980/reviews/?...   

                                               specs         price   
0                                    [Single-player]          4.99  \
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   

   early_access      id         developer        sentiment metascore  
0         False  761140         Kotoshiro              NaN       NaN  
1         False  643980  Secret Level SRL  Mostly Positive       NaN

## Columns: url & reviews_url
We do not need any url for the analysis so we will drop them.

In [35]:
rawGamesDf.drop(columns=['url', 'reviews_url'], inplace=True)

In [36]:
rawGamesDf.head(2)

publisher                                         genres   
0         Kotoshiro  [Action, Casual, Indie, Simulation, Strategy]  \
1  Making Fun, Inc.           [Free to Play, Indie, RPG, Strategy]   

              app_name                title release_date   
0  Lost Summoner Kitty  Lost Summoner Kitty   2018-01-04  \
1            Ironbound            Ironbound   2018-01-04   

                                                tags  discount_price   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49  \
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   

                                               specs         price   
0                                    [Single-player]          4.99  \
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   

   early_access      id         developer        sentiment metascore  
0         False  761140         Kotoshiro              NaN       NaN  
1         False  643980  Secret Level SRL  Mostly Positive       NaN

---

## Duplicates

Analysis if there are duplicates

In [37]:
# Games with duplicate ids, dropna to avoid games with ids = NaN.
rawGamesDf[rawGamesDf.duplicated(keep=False, subset='id')].dropna(subset='id')

publisher    genres                          app_name   
13894  Bethesda Softworks  [Action]  Wolfenstein II: The New Colossus  \
14573  Bethesda Softworks  [Action]  Wolfenstein II: The New Colossus   

                                  title release_date   
13894  Wolfenstein II: The New Colossus   2017-10-26  \
14573  Wolfenstein II: The New Colossus   2017-10-26   

                                                    tags  discount_price   
13894  [Action, FPS, Gore, Violent, Alternate History...             NaN  \
14573  [Action, FPS, Gore, Violent, Alternate History...             NaN   

                                                   specs  price  early_access   
13894  [Single-player, Steam Achievements, Full contr...  59.99         False  \
14573  [Single-player, Steam Achievements, Full contr...  59.99         False   

           id      developer        sentiment metascore  
13894  612880  Machine Games  Mostly Positive        86  
14573  612880  Machine Games  Mostly Positive        86

In [38]:
index_last_dup_to_remove = rawGamesDf[rawGamesDf.duplicated(keep='last', subset='id')].dropna(subset='id').index
index_last_dup_to_remove

Index([13894], dtype='int64')

In [39]:
rawGamesDf.iloc[index_last_dup_to_remove]

publisher    genres                          app_name   
13894  Bethesda Softworks  [Action]  Wolfenstein II: The New Colossus  \

                                  title release_date   
13894  Wolfenstein II: The New Colossus   2017-10-26  \

                                                    tags  discount_price   
13894  [Action, FPS, Gore, Violent, Alternate History...             NaN  \

                                                   specs  price  early_access   
13894  [Single-player, Steam Achievements, Full contr...  59.99         False  \

           id      developer        sentiment metascore  
13894  612880  Machine Games  Mostly Positive        86

In [40]:
rawGamesDf.drop(index=index_last_dup_to_remove, inplace=True)

In [41]:
# Can't use this one yet because of the list in some columns.
# rawGamesDf.drop_duplicates(inplace=True)

---

### Aux Functions

In [42]:
def sort_dic(dic: dict):
    """
    Function made to sort dictionary values for the top 5 rankings.
    """
    dic_sorted = {}
    for y in dic.keys():
        d = dic[y]
        d_sorted = {k: v for k, v in sorted(d.items(), key=lambda item: item[1], reverse=True)}
        dic_sorted[y] = d_sorted
    return dic_sorted

---

## Endpoints, general

Since the data is static and won't change, the most optimal solution could be to prepare a data with every year already processed so the API endpoint can consume it with the least processing in real time.

And since every endpoint to be consumed by the API it needs to know the year, let's process that first.

### Year

We only need year for this study so we will create a new column called year with that information. Extracting year with a pattern of four numbers toghether.

In [43]:
gamesWithYearDf = rawGamesDf

In [44]:
gamesWithYearDf['release_date'].unique()

array(['2018-01-04', '2017-07-24', '2017-12-07', ..., '2016-11-19',
       'January 2018', '2018-10-01'], dtype=object)

In [45]:
# Pattern of regex to extract, 1XXX or 2XXX (valid years).
pattern = '([1-2][0-9][0-9][0-9])'
gamesWithYearDf['year'] = gamesWithYearDf['release_date'].str.extract(pattern, expand = True)

In [46]:
gamesWithYearDf['year'].head()

0    2018
1    2018
2    2017
3    2017
4     NaN
Name: year, dtype: object

In [47]:
gamesWithYearDf['year'].describe()

count     29966
unique       44
top        2017
freq       9594
Name: year, dtype: object

In [48]:
gamesWithYearDf['year'].isna().sum()

2168

Drop "release_date" and drop NaNs from year since they will not give us relevant information on the APIs queries.

In [49]:
gamesWithYearDf = gamesWithYearDf.drop(columns='release_date')

In [50]:
gamesWithYearDf.shape

(32134, 14)

In [51]:
gamesWithYearDf = gamesWithYearDf.dropna(subset=['year'])

In [52]:
gamesWithYearDf.shape

(29966, 14)

Lastly we convert years into int and we can check the years so they are valid years and drop the invalids.

In [53]:
gamesWithYearDf['year'] = pd.to_numeric(gamesWithYearDf['year'])
gamesWithYearDf['year'].info()

<class 'pandas.core.series.Series'>
Index: 29966 entries, 0 to 32133
Series name: year
Non-Null Count  Dtype
--------------  -----
29966 non-null  int64
dtypes: int64(1)
memory usage: 468.2 KB


In [54]:
gamesWithYearDf[gamesWithYearDf['year'] > 2023]

publisher                         genres             app_name   
13428    一次元创作组  [Casual, Indie, Early Access]  Puzzle Sisters Foer  \

                     title                           tags  discount_price   
13428  Puzzle Sisters Foer  [Early Access, Casual, Indie]             NaN  \

                                                   specs price  early_access   
13428  [Single-player, Steam Achievements, Steam Trad...   NaN          True  \

           id developer sentiment metascore  year  
13428  710190    一次元创作组       NaN       NaN  2756

In [55]:
gamesWithYearDf[gamesWithYearDf['year'] < 1970].head()

Empty DataFrame
Columns: [publisher, genres, app_name, title, tags, discount_price, specs, price, early_access, id, developer, sentiment, metascore, year]
Index: []

Let's remove invalide, more than 2023. And older games

In [30]:
gamesWithYearDf = gamesWithYearDf[gamesWithYearDf['year'] <= 2023]

In [29]:
gamesWithYearDf = gamesWithYearDf[gamesWithYearDf['year'] > 1970]

### App_name, Title

Also, study both columns which contane the names of the game, "app_name" and "title".

In [56]:
gamesWithYearDf[['app_name', 'title']].head()

app_name                    title
0      Lost Summoner Kitty      Lost Summoner Kitty
1                Ironbound                Ironbound
2  Real Pool 3D - Poolians  Real Pool 3D - Poolians
3                  弹炸人2222                  弹炸人2222
5    Battle Royale Trainer    Battle Royale Trainer

First let's study and unify app_name and title into one. It only has one NaN each and it is in the same row, so we drop it.

In [57]:
print("Unique:",gamesWithYearDf['app_name'].unique())
print("NaNs:", gamesWithYearDf['app_name'].isna().sum())

Unique: ['Lost Summoner Kitty' 'Ironbound' 'Real Pool 3D - Poolians' ...
 'LOGistICAL: South Africa' 'Russian Roads' 'EXIT 2 - Directions']
NaNs: 1


In [58]:
print("Unique:",gamesWithYearDf['title'].unique())
print("NaNs:", gamesWithYearDf['title'].isna().sum())

Unique: ['Lost Summoner Kitty' 'Ironbound' 'Real Pool 3D - Poolians' ...
 'LOGistICAL: South Africa' 'Russian Roads' 'EXIT 2 - Directions']
NaNs: 1


In [59]:
gamesWithYearDf[gamesWithYearDf['title'].isna()]

publisher           genres app_name title             tags   
2580       NaN  [Action, Indie]      NaN   NaN  [Action, Indie]  \

      discount_price                       specs price  early_access      id   
2580             NaN  [Single-player, Game demo]   NaN         False  317160  \

     developer sentiment metascore  year  
2580       NaN  Positive       NaN  2014

In [60]:
gamesWithYearDf.shape

(29966, 14)

In [61]:
gamesWithYearDf = gamesWithYearDf.dropna(subset=['app_name', 'title'])

In [62]:
gamesWithYearDf.shape

(29965, 14)

Comparing the difference of game names it looks better the characters in "app_name", so we drop "title".

In [63]:
gamesWithYearDf[gamesWithYearDf['app_name'] != gamesWithYearDf['title']].tail()

publisher               genres   
31871  Telltale Games  [Action, Adventure]  \
31872  Telltale Games  [Action, Adventure]   
31873  Telltale Games  [Action, Adventure]   
31898  Strategy First   [Action, Strategy]   
32052         Ubisoft        [Action, RPG]   

                                     app_name   
31871              Sam & Max 105: Reality 2.0  \
31872    Sam & Max 104: Abe Lincoln Must Die!   
31873  Sam & Max 106: Bright Side of the Moon   
31898    Making History: The Calm & the Storm   
32052           Dark Messiah of Might & Magic   

                                            title   
31871              Sam &amp; Max 105: Reality 2.0  \
31872    Sam &amp; Max 104: Abe Lincoln Must Die!   
31873  Sam &amp; Max 106: Bright Side of the Moon   
31898    Making History: The Calm &amp; the Storm   
32052           Dark Messiah of Might &amp; Magic   

                                                    tags  discount_price   
31871                                [Adventure, Action]             NaN  \
31872  [Adventure, Action, Point & Click, Free to Pla...             NaN   
31873                                [Adventure, Action]             NaN   
31898  [Strategy, Action, Turn-Based Strategy, Grand ...             NaN   
32052  [RPG, Action, First-Person, Fantasy, Adventure...             NaN   

                                                   specs  price  early_access   
31871                                    [Single-player]  19.99         False  \
31872                                    [Single-player]  19.99         False   
31873                                    [Single-player]  19.99         False   
31898                      [Single-player, Multi-player]   4.99         False   
32052  [Single-player, Multi-player, Valve Anti-Cheat...   9.99         False   

         id       developer       sentiment metascore  year  
31871  8240  Telltale Games        Positive       NaN  2007  
31872  8230  Telltale Games  8 user reviews       NaN  2007  
31873  8250  Telltale Games        Positive       NaN  2007  
31898  6250      Muzzy Lane   Very Positive        70  2007  
32052  2100  Arkane Studios   Very Positive        72  2006

In [64]:
gamesWithYearDf = gamesWithYearDf.drop(columns='title')

---
### Export JSON for future EDA
Now that we can read the dataset, let's export to a new json file so we can use it in the EDA later.

In [65]:
gamesWithYearDf.to_json("datasets/steam_games_format_ok.json")

---

## Endpoint 1: genero

def genero( Año: str ): Se ingresa un año y devuelve una lista con los 5 géneros más vendidos en el orden correspondiente.

Example:
| Year | Genres |
| -- | - |
| 2018 | {'Indie': 125, 'Action': 75, 'Adventure': 74, 'Casual': 45, 'Simulation': 43} |
| 2016 | {'Indie': 4106, 'Action': 2544, 'Casual': 2141, 'Adventure': 2042, 'Simulation': 1666} |

For this case the only thing that matters to us is "genres" and "year" columns, so we can drop the other ones.


In [66]:
generoEndpointDf = gamesWithYearDf[['genres', 'year']]
generoEndpointDf.head()

genres  year
0      [Action, Casual, Indie, Simulation, Strategy]  2018
1               [Free to Play, Indie, RPG, Strategy]  2018
2  [Casual, Free to Play, Indie, Simulation, Sports]  2017
3                        [Action, Adventure, Casual]  2017
5                    [Action, Adventure, Simulation]  2018

Now we can drop all the columns that have a NaN in any of the two columns because they will bring no information to the scope of the genero function.

In [67]:
generoEndpointDf.describe()

year
count  29965.000000
mean    2014.794727
std        5.532999
min     1970.000000
25%     2014.000000
50%     2016.000000
75%     2017.000000
max     2756.000000

In [68]:
generoEndpointDf.isna().sum()

genres    1234
year         0
dtype: int64

In [69]:
generoEndpointDf = generoEndpointDf.dropna()

In [70]:
generoEndpointDf.isna().sum()

genres    0
year      0
dtype: int64

### Genre

Each game can have various genres, we can see this in the "genres" column because it is filled with lists.

For this reason we need to explode it so we can count all the genres separately.

In [71]:
genresCompleteList = generoEndpointDf.explode(column=['genres'])['genres'].unique()
print(genresCompleteList)
print(genresCompleteList.shape)

['Action' 'Casual' 'Indie' 'Simulation' 'Strategy' 'Free to Play' 'RPG'
 'Sports' 'Adventure' 'Racing' 'Massively Multiplayer' 'Early Access'
 'Animation &amp; Modeling' 'Video Production' 'Web Publishing'
 'Education' 'Software Training' 'Utilities' 'Design &amp; Illustration'
 'Audio Production' 'Photo Editing' 'Accounting']
(22,)


In [72]:
generoEndpointDf.shape

(28731, 2)

In [73]:
generoEndpointDf = generoEndpointDf.explode('genres')
generoEndpointDf.head()

genres  year
0      Action  2018
0      Casual  2018
0       Indie  2018
0  Simulation  2018
0    Strategy  2018

In [74]:
generoEndpointDf.shape

(71213, 2)

Now we can make a Groupby to get all the genres of each year. Then unstack it and have a new Data Frame with the data we desire.

In [75]:
generoEndpointDf2 = generoEndpointDf.groupby(by=['year', 'genres'])['genres'].count()
generoEndpointDf2.tail()

year  genres      
2021  Indie           1
      RPG             1
2756  Casual          1
      Early Access    1
      Indie           1
Name: genres, dtype: int64

In [76]:
generoEndpointDf3 = generoEndpointDf2.unstack(level=1, fill_value=0)
generoEndpointDf3.head()

genres  Accounting  Action  Adventure  Animation &amp; Modeling   
year                                                              
1983             0       1          1                         0  \
1984             0       1          1                         0   
1985             0       0          0                         0   
1986             0       0          1                         0   
1987             0       0          3                         0   

genres  Audio Production  Casual  Design &amp; Illustration  Early Access   
year                                                                        
1983                   0       1                          0             0  \
1984                   0       2                          0             0   
1985                   0       0                          0             0   
1986                   0       1                          0             0   
1987                   0       0                          0             0   

genres  Education  Free to Play  ...  Photo Editing  RPG  Racing  Simulation   
year                             ...                                           
1983            0             0  ...              0    0       0           0  \
1984            0             0  ...              0    0       0           0   
1985            0             0  ...              0    0       0           1   
1986            0             0  ...              0    1       0           1   
1987            0             0  ...              0    0       0           2   

genres  Software Training  Sports  Strategy  Utilities  Video Production   
year                                                                       
1983                    0       0         0          0                 0  \
1984                    0       0         0          0                 0   
1985                    0       0         0          0                 0   
1986                    0       0         0          0                 0   
1987                    0       0         0          0                 0   

genres  Web Publishing  
year                    
1983                 0  
1984                 0  
1985                 0  
1986                 0  
1987                 0  

[5 rows x 22 columns]

In [77]:
# Now we have exactly what we need.
generoEndpointDict = generoEndpointDf3.to_dict(orient='index')
generoEndpointDict

{1983: {'Accounting': 0,
  'Action': 1,
  'Adventure': 1,
  'Animation &amp; Modeling': 0,
  'Audio Production': 0,
  'Casual': 1,
  'Design &amp; Illustration': 0,
  'Early Access': 0,
  'Education': 0,
  'Free to Play': 0,
  'Indie': 0,
  'Massively Multiplayer': 0,
  'Photo Editing': 0,
  'RPG': 0,
  'Racing': 0,
  'Simulation': 0,
  'Software Training': 0,
  'Sports': 0,
  'Strategy': 0,
  'Utilities': 0,
  'Video Production': 0,
  'Web Publishing': 0},
 1984: {'Accounting': 0,
  'Action': 1,
  'Adventure': 1,
  'Animation &amp; Modeling': 0,
  'Audio Production': 0,
  'Casual': 2,
  'Design &amp; Illustration': 0,
  'Early Access': 0,
  'Education': 0,
  'Free to Play': 0,
  'Indie': 1,
  'Massively Multiplayer': 0,
  'Photo Editing': 0,
  'RPG': 0,
  'Racing': 0,
  'Simulation': 0,
  'Software Training': 0,
  'Sports': 0,
  'Strategy': 0,
  'Utilities': 0,
  'Video Production': 0,
  'Web Publishing': 0},
 1985: {'Accounting': 0,
  'Action': 0,
  'Adventure': 0,
  'Animation &amp;

Sort by values.

In [78]:
generoEndpointDict_sorted = sort_dic(generoEndpointDict)
generoEndpointDict_sorted

{1983: {'Action': 1,
  'Adventure': 1,
  'Casual': 1,
  'Accounting': 0,
  'Animation &amp; Modeling': 0,
  'Audio Production': 0,
  'Design &amp; Illustration': 0,
  'Early Access': 0,
  'Education': 0,
  'Free to Play': 0,
  'Indie': 0,
  'Massively Multiplayer': 0,
  'Photo Editing': 0,
  'RPG': 0,
  'Racing': 0,
  'Simulation': 0,
  'Software Training': 0,
  'Sports': 0,
  'Strategy': 0,
  'Utilities': 0,
  'Video Production': 0,
  'Web Publishing': 0},
 1984: {'Casual': 2,
  'Action': 1,
  'Adventure': 1,
  'Indie': 1,
  'Accounting': 0,
  'Animation &amp; Modeling': 0,
  'Audio Production': 0,
  'Design &amp; Illustration': 0,
  'Early Access': 0,
  'Education': 0,
  'Free to Play': 0,
  'Massively Multiplayer': 0,
  'Photo Editing': 0,
  'RPG': 0,
  'Racing': 0,
  'Simulation': 0,
  'Software Training': 0,
  'Sports': 0,
  'Strategy': 0,
  'Utilities': 0,
  'Video Production': 0,
  'Web Publishing': 0},
 1985: {'Simulation': 1,
  'Accounting': 0,
  'Action': 0,
  'Adventure': 0,

In [79]:
with open("datasets/steam_games_endpoint_1_genero.json", "w") as outfile:
    json.dump(generoEndpointDict_sorted, outfile)

**TEST**

In [80]:
def genero(Year: str): 
    """
    Se ingresa un año y devuelve un diccionario con los 5 géneros más vendidos en el orden correspondiente.
    """

    # Read data
    with open("datasets/steam_games_endpoint_1_genero.json") as json_file:
        genres_dict = json.load(json_file)
    # Since data is already sorted, get the first 5 as top 5.
    top5 = dict(list(genres_dict[Year].items())[0:5])
    return top5

genero('2016')

{'Indie': 4106,
 'Action': 2544,
 'Casual': 2141,
 'Adventure': 2042,
 'Simulation': 1666}

---

## Endpoint 2: juegos

def juegos( Año: str ): Se ingresa un año y devuelve una lista con los juegos lanzados en el año.

Example:
| Year | Games |
| -- | - |
| 2018 | ['juego1', 'juego2', ... ] |
| 2016 | ['juegoX', 'juegoY', ... ] |

For this case the only thing that matters to us is "app_name" and "year" columns, so we can drop the other ones.

### Games

Each game has the name on column "app_name". Let's prepare the json.

In [81]:
juegosEndpointDf = gamesWithYearDf[['app_name', 'year']]
juegosEndpointDf.head()

app_name  year
0      Lost Summoner Kitty  2018
1                Ironbound  2018
2  Real Pool 3D - Poolians  2017
3                  弹炸人2222  2017
5    Battle Royale Trainer  2018

In [82]:
juegosEndpointDf.shape

(29965, 2)

Now we can make a Groupby to get all the genres of each year. Then unstack it and have a new Data Frame with the data we desire.

In [84]:
juegosEndpointDf2 = juegosEndpointDf.groupby(by=['year']).apply(lambda df: list(df['app_name']))

juegosEndpointDf2.tail(20)

year
2002    [Arx Fatalis, The Sum of All Fears, Freedom Fo...
2003    [Ghost Master®, Railroad Tycoon 3, BloodRayne,...
2004    [Rome: Total War™ - Collection, Manhunt, Far C...
2005    [Advent Rising, FlatOut, Act of War: Direct Ac...
2006    [Disciples II: Gallean's Return, Disciples II:...
2007    [Lost Planet™: Extreme Condition, RIP - Trilog...
2008    [Command & Conquer: Red Alert 3, Conflict: Den...
2009    [PRR Wagon Pack 01, Class 421 London South Eas...
2010    [Avencast: Rise of the Mage, Pirates, Vikings,...
2011    [Mafia II, Lara Croft GoL: Raziel and Kain Cha...
2012    [Hard Reset Extended Edition, Blackwell Unboun...
2013    [RIFT, Trine 2: Complete Story, Angelica Weave...
2014    [Imagine Earth, BeatBlasters III, Blood of the...
2015    [Assault Android Cactus, America's Army: Provi...
2016    [Fallen Mage, Of Guards And Thieves, The Inter...
2017    [Real Pool 3D - Poolians, 弹炸人2222, RC Plane 3 ...
2018    [Lost Summoner Kitty, Ironbound, Battle Royale...
2019    [

In [85]:
juegosEndpointDict = dict(juegosEndpointDf2)

In [86]:
with open("datasets/steam_games_endpoint_2_juegos.json", "w") as outfile:
    json.dump(juegosEndpointDict, outfile)

**TEST**

In [87]:
def juegos( Year: str ):
    """
    Se ingresa un año y devuelve un diccionario con los juegos lanzados en el año.
    """
    # Read data
    with open("datasets/steam_games_endpoint_2_juegos.json") as json_file:
        juegos_dict = json.load(json_file)
    return juegos_dict[Year]

juegos('2015')

['Assault Android Cactus',
 "America's Army: Proving Grounds",
 'The Tower',
 'Dimmdrive :: Gaming Ramdrive @ 10,000+ MB/s',
 'Fake Colours',
 'WolfWars',
 'Dr.Green',
 'Vintage Year',
 'The Race for the White House',
 'X-Plane 10 AddOn - Carenado - C185F Skywagon',
 'Rocksmith® 2014 – Linkin Park - “What I’ve Done”',
 'Rocksmith® 2014 – Linkin Park Song Pack',
 'Dreaming Sarah OST',
 'Antisquad: Tasks near the coast of Somalia. Tactics DLC',
 'Castle',
 'Antisquad: Tasks in North Korea. Tactics DLC',
 'ATLAS レジェンドパック',
 'Ostrich Island',
 'Adventures of Robinson Crusoe',
 'Mechs & Mercs: Black Talons',
 'Urja',
 'Teddy Floppy Ear - Mountain Adventure',
 'Su-27: The Ultimate Argument Campaign',
 'Squirreltopia',
 'Lara Croft and the Temple of Osiris - Icy Death Pack',
 'Lara Croft and the Temple of Osiris - Legend Pack',
 'Lara Croft and the Temple of Osiris - Deus Ex Pack',
 'Disorder',
 'Lara Croft and the Temple of Osiris - Hitman Pack',
 'Avernum 2: Crystal Souls',
 'Gunslugs 2',
 

---

## Endpoint 3: specs

def specs( Año: str ): Se ingresa un año y devuelve una lista con los 5 specs que más se repiten en el mismo en el orden correspondiente.

Example:
| Year | Specs |
| -- | - |
| 2018 | {'Indie': 125, 'Action': 75, 'Adventure': 74, 'Casual': 45, 'Simulation': 43} |
| 2016 | {'Indie': 4106, 'Action': 2544, 'Casual': 2141, 'Adventure': 2042, 'Simulation': 1666} |

For this case the only thing that matters to us is "specs" and "year" columns, so we can drop the other ones.

In [88]:
specsEndpointDf = gamesWithYearDf[['specs', 'year']]
specsEndpointDf.head()

specs  year
0                                    [Single-player]  2018
1  [Single-player, Multi-player, Online Multi-Pla...  2018
2  [Single-player, Multi-player, Online Multi-Pla...  2017
3                                    [Single-player]  2017
5                [Single-player, Steam Achievements]  2018

Now we can drop all the columns that have a NaN in any of the two columns because they will bring no information to the scope of the genero function.

In [89]:
specsEndpointDf.describe()

year
count  29965.000000
mean    2014.794727
std        5.532999
min     1970.000000
25%     2014.000000
50%     2016.000000
75%     2017.000000
max     2756.000000

In [90]:
specsEndpointDf.isna().sum()

specs    669
year       0
dtype: int64

In [91]:
specsEndpointDf = specsEndpointDf.dropna()

In [92]:
specsEndpointDf.isna().sum()

specs    0
year     0
dtype: int64

### Specs

Similarly to genres, each game can have various specs, we can see this in the "specs" column because it is filled with lists.

For this reason we need to explode it so we can count all the specs separately.

In [93]:
specsCompleteList = specsEndpointDf.explode(column=['specs'])['specs'].unique()
print(specsCompleteList)
print(specsCompleteList.shape)

['Single-player' 'Multi-player' 'Online Multi-Player'
 'Cross-Platform Multiplayer' 'Steam Achievements' 'Steam Trading Cards'
 'In-App Purchases' 'Stats' 'Downloadable Content'
 'Full controller support' 'Steam Cloud' 'Steam Leaderboards'
 'Partial Controller Support' 'Local Co-op' 'Shared/Split Screen'
 'Valve Anti-Cheat enabled' 'Local Multi-Player'
 'Steam Turn Notifications' 'MMO' 'Co-op' 'Captions available'
 'Steam Workshop' 'Includes level editor' 'Mods' 'Mods (require HL2)'
 'Online Co-op' 'Game demo' 'Includes Source SDK' 'Commentary available'
 'SteamVR Collectibles' 'Mods (require HL1)']
(31,)


In [94]:
specsEndpointDf.shape

(29296, 2)

In [95]:
specsEndpointDf = specsEndpointDf.explode('specs')
specsEndpointDf.head()

specs  year
0               Single-player  2018
1               Single-player  2018
1                Multi-player  2018
1         Online Multi-Player  2018
1  Cross-Platform Multiplayer  2018

In [96]:
specsEndpointDf.shape

(128231, 2)

Now we can make a Groupby to get all the specs of each year. Then unstack it and have a new Data Frame with the data we desire.

In [97]:
specsEndpointDf2 = specsEndpointDf.groupby(by=['year', 'specs'])['specs'].count()
specsEndpointDf2.tail()

year  specs              
2756  Single-player          1
      Steam Achievements     1
      Steam Cloud            1
      Steam Trading Cards    1
      Steam Workshop         1
Name: specs, dtype: int64

In [98]:
specsEndpointDf3 = specsEndpointDf2.unstack(level=1, fill_value=0)
specsEndpointDf3.head()

specs  Captions available  Co-op  Commentary available   
year                                                     
1970                    2      0                     0  \
1975                    1      0                     0   
1980                    1      0                     0   
1981                    2      0                     0   
1982                    3      0                     0   

specs  Cross-Platform Multiplayer  Downloadable Content   
year                                                      
1970                            0                     0  \
1975                            0                     0   
1980                            0                     0   
1981                            0                     0   
1982                            0                     0   

specs  Full controller support  Game demo  In-App Purchases   
year                                                          
1970                         0          0                 0  \
1975                         0          0                 0   
1980                         0          0                 0   
1981                         0          0                 0   
1982                         0          0                 0   

specs  Includes Source SDK  Includes level editor  ...  Single-player  Stats   
year                                               ...                         
1970                     0                      0  ...              0      0  \
1975                     0                      0  ...              0      0   
1980                     0                      0  ...              0      0   
1981                     0                      0  ...              0      0   
1982                     0                      0  ...              0      0   

specs  Steam Achievements  Steam Cloud  Steam Leaderboards   
year                                                         
1970                    0            0                   0  \
1975                    0            0                   0   
1980                    0            0                   0   
1981                    0            0                   0   
1982                    0            0                   0   

specs  Steam Trading Cards  Steam Turn Notifications  Steam Workshop   
year                                                                   
1970                     0                         0               0  \
1975                     0                         0               0   
1980                     0                         0               0   
1981                     0                         0               0   
1982                     0                         0               0   

specs  SteamVR Collectibles  Valve Anti-Cheat enabled  
year                                                   
1970                      0                         0  
1975                      0                         0  
1980                      0                         0  
1981                      0                         0  
1982                      0                         0  

[5 rows x 31 columns]

In [99]:
# Now we have exactly what we need.
specsEndpointDict = specsEndpointDf3.to_dict(orient='index')
specsEndpointDict

{1970: {'Captions available': 2,
  'Co-op': 0,
  'Commentary available': 0,
  'Cross-Platform Multiplayer': 0,
  'Downloadable Content': 0,
  'Full controller support': 0,
  'Game demo': 0,
  'In-App Purchases': 0,
  'Includes Source SDK': 0,
  'Includes level editor': 0,
  'Local Co-op': 0,
  'Local Multi-Player': 0,
  'MMO': 0,
  'Mods': 0,
  'Mods (require HL1)': 0,
  'Mods (require HL2)': 0,
  'Multi-player': 0,
  'Online Co-op': 0,
  'Online Multi-Player': 0,
  'Partial Controller Support': 0,
  'Shared/Split Screen': 0,
  'Single-player': 0,
  'Stats': 0,
  'Steam Achievements': 0,
  'Steam Cloud': 0,
  'Steam Leaderboards': 0,
  'Steam Trading Cards': 0,
  'Steam Turn Notifications': 0,
  'Steam Workshop': 0,
  'SteamVR Collectibles': 0,
  'Valve Anti-Cheat enabled': 0},
 1975: {'Captions available': 1,
  'Co-op': 0,
  'Commentary available': 0,
  'Cross-Platform Multiplayer': 0,
  'Downloadable Content': 0,
  'Full controller support': 0,
  'Game demo': 0,
  'In-App Purchases':

Sort by values.

In [100]:
specsEndpointDict_sorted = {}
for y in specsEndpointDict.keys():
    specs = specsEndpointDict[y]
    specs_sorted = {k: v for k, v in sorted(specs.items(), key=lambda item: item[1], reverse=True)}
    specsEndpointDict_sorted[y] = specs_sorted

specsEndpointDict_sorted

{1970: {'Captions available': 2,
  'Co-op': 0,
  'Commentary available': 0,
  'Cross-Platform Multiplayer': 0,
  'Downloadable Content': 0,
  'Full controller support': 0,
  'Game demo': 0,
  'In-App Purchases': 0,
  'Includes Source SDK': 0,
  'Includes level editor': 0,
  'Local Co-op': 0,
  'Local Multi-Player': 0,
  'MMO': 0,
  'Mods': 0,
  'Mods (require HL1)': 0,
  'Mods (require HL2)': 0,
  'Multi-player': 0,
  'Online Co-op': 0,
  'Online Multi-Player': 0,
  'Partial Controller Support': 0,
  'Shared/Split Screen': 0,
  'Single-player': 0,
  'Stats': 0,
  'Steam Achievements': 0,
  'Steam Cloud': 0,
  'Steam Leaderboards': 0,
  'Steam Trading Cards': 0,
  'Steam Turn Notifications': 0,
  'Steam Workshop': 0,
  'SteamVR Collectibles': 0,
  'Valve Anti-Cheat enabled': 0},
 1975: {'Captions available': 1,
  'Co-op': 0,
  'Commentary available': 0,
  'Cross-Platform Multiplayer': 0,
  'Downloadable Content': 0,
  'Full controller support': 0,
  'Game demo': 0,
  'In-App Purchases':

In [101]:
with open("datasets/steam_games_endpoint_3_specs.json", "w") as outfile:
    json.dump(specsEndpointDict_sorted, outfile)

**TEST**

In [102]:
def specs( Year: str ):
    """
    Se ingresa un año y devuelve un diccionario con los 5 specs que más se repiten en el mismo en el orden correspondiente.
    """
    # Read data
    with open("datasets/steam_games_endpoint_3_specs.json") as json_file:
        specs_dict = json.load(json_file)

    # TODO Add validation

    # Since data is already sorted, get the first 5 as top 5.
    top5 = dict(list(specs_dict[Year].items())[0:5])
    return top5

specs('2018')

{'Single-player': 147,
 'Steam Achievements': 68,
 'Full controller support': 44,
 'Steam Cloud': 30,
 'Partial Controller Support': 29}

---

## Endpoint 4: earlyacces

def earlyaccess( Año: str ): Cantidad de juegos lanzados en un año con early access.

Example:
| Year | earlyacces |
| -- | - |
| 2018 | {125} |
| 2016 | {1234} |

For this case the only thing that matters to us is "earlyacces" and "year" columns, so we can drop the other ones.

In [103]:
earlyaccessEndpointDf = gamesWithYearDf[['early_access', 'year']]
earlyaccessEndpointDf.head()

early_access  year
0         False  2018
1         False  2018
2         False  2017
3         False  2017
5         False  2018

Now we can drop all the columns that have a NaN in any of the two columns because they will bring no information to the scope of the genero function.

In [104]:
earlyaccessEndpointDf['early_access'].describe()

count     29965
unique        2
top       False
freq      28519
Name: early_access, dtype: object

In [105]:
earlyaccessEndpointDf.isna().sum()

early_access    0
year            0
dtype: int64

### Early Access

Just need to group by by ear and count amount of early acces games.

In [106]:
earlyaccessEndpointDf2 = earlyaccessEndpointDf.groupby(by=['year']).sum()
earlyaccessEndpointDf2.tail()

early_access
year              
2017           733
2018            33
2019             1
2021             0
2756             1

In [107]:
# Now we have exactly what we need.
earlyaccessEndpointDict = earlyaccessEndpointDf2.to_dict()
earlyaccessEndpointDict

{'early_access': {1970: 0,
  1975: 0,
  1980: 0,
  1981: 0,
  1982: 0,
  1983: 0,
  1984: 0,
  1985: 0,
  1986: 0,
  1987: 0,
  1988: 0,
  1989: 0,
  1990: 0,
  1991: 0,
  1992: 0,
  1993: 0,
  1994: 0,
  1995: 0,
  1996: 0,
  1997: 0,
  1998: 0,
  1999: 0,
  2000: 0,
  2001: 1,
  2002: 0,
  2003: 0,
  2004: 0,
  2005: 0,
  2006: 0,
  2007: 0,
  2008: 0,
  2009: 0,
  2010: 0,
  2011: 0,
  2012: 2,
  2013: 10,
  2014: 78,
  2015: 224,
  2016: 363,
  2017: 733,
  2018: 33,
  2019: 1,
  2021: 0,
  2756: 1}}

In [108]:
with open("datasets/steam_games_endpoint_4_earlyaccess.json", "w") as outfile:
    json.dump(earlyaccessEndpointDict, outfile)

**TEST**

In [110]:
def earlyacces( Year: str ):
    """
    Cantidad de juegos lanzados en un año con early access.
    """
    # Read data
    with open("datasets/steam_games_endpoint_4_earlyaccess.json") as json_file:
        earlyaccess_dict = json.load(json_file)

    # TODO Add validation, Top5 si teine al menos 5

    return {"Año": Year, "Early Access": earlyaccess_dict['early_access'][Year]}


earlyacces('2015')

{'Año': '2015', 'Early Access': 224}

---

## Endpoint 5: sentiment

def sentiment( Año: str ): Según el año de lanzamiento, se devuelve una lista con la cantidad de registros que se encuentren categorizados con un análisis de sentimiento.

Ejemplo de retorno: {Mixed = 182, Very Positive = 120, Positive = 278}

For this case the only thing that matters to us is "sentiment" and "year" columns, so we can drop the other ones.

In [111]:
sentimentEndpointDf = gamesWithYearDf[['sentiment', 'year']]
sentimentEndpointDf.head()

sentiment  year
0              NaN  2018
1  Mostly Positive  2018
2  Mostly Positive  2017
3              NaN  2017
5            Mixed  2018

Now we can drop all the columns that have a NaN in any of the two columns because they will bring no information to the scope of the genero function.

In [112]:
sentimentEndpointDf.shape

(29965, 2)

In [113]:
sentimentEndpointDf.isna().sum()

sentiment    6811
year            0
dtype: int64

In [114]:
sentimentEndpointDf = sentimentEndpointDf.dropna()

In [115]:
sentimentEndpointDf.isna().sum()

sentiment    0
year         0
dtype: int64

In [116]:
sentimentEndpointDf.shape

(23154, 2)

### Sentiment

In this column we have various "sentiment" attributes. In the case that there are too few reviews it puts the amount of reviews instead of sentiment.

We could remove all that have "% user reviews"

In [117]:
sentimentEndpointDf['sentiment'].value_counts()

sentiment
Mixed                      3861
Very Positive              3659
Positive                   2965
Mostly Positive            2563
1 user reviews             2348
2 user reviews             1633
3 user reviews             1127
4 user reviews              882
5 user reviews              771
Mostly Negative             771
6 user reviews              673
7 user reviews              548
8 user reviews              475
9 user reviews              433
Overwhelmingly Positive     293
Negative                    116
Very Negative                29
Overwhelmingly Negative       7
Name: count, dtype: int64

In [119]:
# Pattern of regex to extract "% user reviews".
pattern_to_del = '[0-9] user reviews'

filter = sentimentEndpointDf['sentiment'].str.contains(pattern_to_del)

sentimentEndpointDf2 = sentimentEndpointDf[~filter]

In [120]:
sentimentEndpointDf2['sentiment'].value_counts()

sentiment
Mixed                      3861
Very Positive              3659
Positive                   2965
Mostly Positive            2563
Mostly Negative             771
Overwhelmingly Positive     293
Negative                    116
Very Negative                29
Overwhelmingly Negative       7
Name: count, dtype: int64

Now we can make a Groupby to get all the genres of each year. Then unstack it and have a new Data Frame with the data we desire.

In [121]:
sentimentEndpointDf3 = sentimentEndpointDf2.groupby(by=['year', 'sentiment'])['sentiment'].count()
sentimentEndpointDf3.tail()

year  sentiment      
2018  Mixed              6
      Mostly Negative    1
      Mostly Positive    3
      Very Positive      3
2021  Positive           1
Name: sentiment, dtype: int64

In [122]:
sentimentEndpointDf4 = sentimentEndpointDf3.unstack(level=1, fill_value=0)
sentimentEndpointDf4.head()

sentiment  Mixed  Mostly Negative  Mostly Positive  Negative   
year                                                           
1982           0                0                0         0  \
1983           0                0                0         0   
1984           0                0                0         0   
1986           1                0                0         0   
1987           2                0                1         0   

sentiment  Overwhelmingly Negative  Overwhelmingly Positive  Positive   
year                                                                    
1982                             0                        0         1  \
1983                             0                        0         0   
1984                             0                        0         1   
1986                             0                        0         0   
1987                             0                        0         1   

sentiment  Very Negative  Very Positive  
year                                     
1982                   0              0  
1983                   0              1  
1984                   0              1  
1986                   0              0  
1987                   0              0

In [123]:
# Now we have exactly what we need.
sentimentEndpointDict = sentimentEndpointDf4.to_dict(orient='index')
sentimentEndpointDict

{1982: {'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 1,
  'Very Negative': 0,
  'Very Positive': 0},
 1983: {'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 0,
  'Very Negative': 0,
  'Very Positive': 1},
 1984: {'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 1,
  'Very Negative': 0,
  'Very Positive': 1},
 1986: {'Mixed': 1,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 0,
  'Very Negative': 0,
  'Very Positive': 0},
 1987: {'Mixed': 2,
  'Mostly Negative': 0,
  'Mostly Positive': 1,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Posi

Sort by values.

In [125]:
sentimentEndpointDict_sorted = sort_dic(sentimentEndpointDict)
sentimentEndpointDict_sorted

{1982: {'Positive': 1,
  'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Very Negative': 0,
  'Very Positive': 0},
 1983: {'Very Positive': 1,
  'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 0,
  'Very Negative': 0},
 1984: {'Positive': 1,
  'Very Positive': 1,
  'Mixed': 0,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Very Negative': 0},
 1986: {'Mixed': 1,
  'Mostly Negative': 0,
  'Mostly Positive': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Positive': 0,
  'Positive': 0,
  'Very Negative': 0,
  'Very Positive': 0},
 1987: {'Mixed': 2,
  'Mostly Positive': 1,
  'Positive': 1,
  'Mostly Negative': 0,
  'Negative': 0,
  'Overwhelmingly Negative': 0,
  'Overwhelmingly Posi

In [126]:
with open("datasets/steam_games_endpoint_5_sentiment.json", "w") as outfile:
    json.dump(sentimentEndpointDict_sorted, outfile)

**TEST**

In [127]:
def sentiment( Year: str ):
    """
    Según el año de lanzamiento, se devuelve una lista con la cantidad de registros que se encuentren categorizados con un análisis de sentimiento.

    Ejemplo de retorno: {Mixed = 182, Very Positive = 120, Positive = 278}
    """
    # Read data
    with open("datasets/steam_games_endpoint_5_sentiment.json") as json_file:
        sentiment_dict = json.load(json_file)

    # TODO Add validation

    return {"Año": Year, "Sentiment": sentiment_dict[Year]}

sentiment('2016')

{'Año': '2016',
 'Sentiment': {'Mixed': 941,
  'Very Positive': 767,
  'Positive': 695,
  'Mostly Positive': 562,
  'Mostly Negative': 215,
  'Overwhelmingly Positive': 49,
  'Negative': 23,
  'Very Negative': 6,
  'Overwhelmingly Negative': 1}}

---

## Endpoint 6: metascore

def metascore( Año: str ): Top 5 juegos según año con mayor metascore.

Example:
| Year | Genres |
| -- | - |
| 2018 | {'Game7': 85, 'Game91': 75, 'Game1232': 74, ...} |
| 2017 | {'Game56': 98, 'Game6': 84, 'Game32': 82, ...} |

For this case the only thing that matters to us is "metascore", "app_name" and "year" columns, so we can drop the other ones.

In [128]:
metascoreEndpointDf = gamesWithYearDf[['app_name', 'metascore', 'year']]
metascoreEndpointDf.head()

app_name metascore  year
0      Lost Summoner Kitty       NaN  2018
1                Ironbound       NaN  2018
2  Real Pool 3D - Poolians       NaN  2017
3                  弹炸人2222       NaN  2017
5    Battle Royale Trainer       NaN  2018

Now we can drop all the columns that have a NaN in any of the two columns because they will bring no information to the scope of the metascore function.

In [129]:
metascoreEndpointDf.shape

(29965, 3)

In [130]:
metascoreEndpointDf.isna().sum()

app_name         0
metascore    27343
year             0
dtype: int64

In [131]:
metascoreEndpointDf = metascoreEndpointDf.dropna()

In [132]:
metascoreEndpointDf.isna().sum()

app_name     0
metascore    0
year         0
dtype: int64

In [133]:
metascoreEndpointDf.shape

(2622, 3)

### Metascore

In [136]:
metascoreEndpointDf.shape

(2622, 3)

In [137]:
metascoreEndpointDf.head()

app_name metascore  year
28                        Half-Life        96  1998
39   Disciples II: Gallean's Return        84  2006
40  Disciples II: Rise of the Elves        80  2006
41           The Ship: Murder Party        76  2006
55                    Advent Rising        70  2005

In [138]:
metascoreEndpointDf['metascore'].unique()

array([96, 84, 80, 76, 70, 'NA', 69, 81, 75, 72, 66, 67, 77, 91, 89, 83,
       61, 88, 65, 94, 57, 86, 87, 92, 79, 82, 58, 74, 85, 90, 68, 71, 60,
       73, 59, 64, 54, 53, 78, 51, 44, 63, 38, 56, 49, 52, 62, 93, 48, 34,
       95, 43, 55, 24, 46, 41, 20, 39, 45, 35, 47, 40, 36, 50, 32, 37, 33,
       42, 27, 29, 30], dtype=object)

Convert to int, remove 'NA' to metascore because it does not give us any information.

In [139]:
metascoreEndpointDf = metascoreEndpointDf[metascoreEndpointDf['metascore'] != 'NA']

In [140]:
metascoreEndpointDf.shape

(2552, 3)

In [141]:
metascoreEndpointDf['metascore'].unique()

array([96, 84, 80, 76, 70, 69, 81, 75, 72, 66, 67, 77, 91, 89, 83, 61, 88,
       65, 94, 57, 86, 87, 92, 79, 82, 58, 74, 85, 90, 68, 71, 60, 73, 59,
       64, 54, 53, 78, 51, 44, 63, 38, 56, 49, 52, 62, 93, 48, 34, 95, 43,
       55, 24, 46, 41, 20, 39, 45, 35, 47, 40, 36, 50, 32, 37, 33, 42, 27,
       29, 30], dtype=object)

Now we can make a Groupby to get all the genres of each year. Then unstack it and have a new Data Frame with the data we desire.

In [142]:
metascoreEndpointDf2 = metascoreEndpointDf.groupby(by=['year']).apply(lambda df: dict(zip(df['app_name'], df['metascore'])))
metascoreEndpointDf2.head(20)

year
1992                               {'Reservoir Dogs': 50}
1993                                     {'realMYST': 66}
1994                                      {'DOOM II': 83}
1996    {'Harvester': 53, 'Star Control: Kessari Quadr...
1997    {'POSTAL': 56, 'The Last Express Gold Edition'...
1998    {'Half-Life': 96, 'Railroad Tycoon II Platinum...
1999    {'X: Beyond the Frontier': 67, 'Gabriel Knight...
2000    {'The Longest Journey': 91, 'Unreal Tournament...
2001    {'Return to Castle Wolfenstein': 88, 'Max Payn...
2002    {'Arx Fatalis': 77, 'The Sum of All Fears': 72...
2003    {'Ghost Master®': 81, 'Railroad Tycoon 3': 80,...
2004    {'Rome: Total War™ - Collection': 92, 'Manhunt...
2005    {'Advent Rising': 70, 'FlatOut': 72, 'Act of W...
2006    {'Disciples II: Gallean's Return': 84, 'Discip...
2007    {'Lost Planet™: Extreme Condition': 66, 'Tomb ...
2008    {'Command & Conquer: Red Alert 3': 82, 'Confli...
2009    {'Saints Row 2': 72, 'Mirror's Edge™': 81, 'F....
2010    {

In [143]:
# Now we have exactly what we need.
metascoreEndpointDict = metascoreEndpointDf2.to_dict()
metascoreEndpointDict

{1992: {'Reservoir Dogs': 50},
 1993: {'realMYST': 66},
 1994: {'DOOM II': 83},
 1996: {'Harvester': 53,
  'Star Control: Kessari Quadrant': 89,
  'FROM DUSK TILL DAWN': 58,
  'Lords of the Realm II': 77},
 1997: {'POSTAL': 56,
  'The Last Express Gold Edition': 82,
  'Total Annihilation': 86,
  'Fallout: A Post Nuclear Role Playing Game': 89,
  'Tony Tough and the Night of Roasted Moths': 68,
  'Shadow Warrior (Classic)': 73,
  'Riven: The Sequel to MYST': 83,
  'STAR WARS™ Jedi Knight: Dark Forces II': 91},
 1998: {'Half-Life': 96,
  'Railroad Tycoon II Platinum': 89,
  'Fallout 2: A Post Nuclear Role Playing Game': 86,
  'Shogo: Mobile Armor Division': 88,
  'Falcon 4.0': 85,
  'Blood: One Unit Whole Blood': 82,
  'Blood II: The Chosen + Expansion': 72,
  'Thief™ Gold': 92},
 1999: {'X: Beyond the Frontier': 67,
  'Gabriel Knight® 3: Blood of the Sacred, Blood of the Damned': 80,
  'Army Men II': 67,
  'Septerra Core': 72,
  'System Shock 2': 92,
  'Dawn of Discovery™: Venice': 78},

Sort by values.

In [144]:
metascoreEndpointDict_sorted = sort_dic(metascoreEndpointDict)
metascoreEndpointDict_sorted

{1992: {'Reservoir Dogs': 50},
 1993: {'realMYST': 66},
 1994: {'DOOM II': 83},
 1996: {'Star Control: Kessari Quadrant': 89,
  'Lords of the Realm II': 77,
  'FROM DUSK TILL DAWN': 58,
  'Harvester': 53},
 1997: {'STAR WARS™ Jedi Knight: Dark Forces II': 91,
  'Fallout: A Post Nuclear Role Playing Game': 89,
  'Total Annihilation': 86,
  'Riven: The Sequel to MYST': 83,
  'The Last Express Gold Edition': 82,
  'Shadow Warrior (Classic)': 73,
  'Tony Tough and the Night of Roasted Moths': 68,
  'POSTAL': 56},
 1998: {'Half-Life': 96,
  'Thief™ Gold': 92,
  'Railroad Tycoon II Platinum': 89,
  'Shogo: Mobile Armor Division': 88,
  'Fallout 2: A Post Nuclear Role Playing Game': 86,
  'Falcon 4.0': 85,
  'Blood: One Unit Whole Blood': 82,
  'Blood II: The Chosen + Expansion': 72},
 1999: {'System Shock 2': 92,
  'Gabriel Knight® 3: Blood of the Sacred, Blood of the Damned': 80,
  'Dawn of Discovery™: Venice': 78,
  'Septerra Core': 72,
  'X: Beyond the Frontier': 67,
  'Army Men II': 67},

In [145]:
with open("datasets/steam_games_endpoint_6_metascore.json", "w") as outfile:
    json.dump(metascoreEndpointDict_sorted, outfile)

**TEST**

In [147]:
def metascore( Year: str ):
    """
    Top 5 juegos según año con mayor metascore.
    """
    # Read data
    with open("datasets/steam_games_endpoint_6_metascore.json") as json_file:
        metascore_dict = json.load(json_file)

    # TODO Add validation, ademas de cantidad menor 5.

    # Since data is already sorted, get the first 5 as top 5.
    top5 = dict(list(metascore_dict[Year].items())[0:5])
    return top5

metascore('2015')

{'Grand Theft Auto V': 96,
 'Divinity: Original Sin - Enhanced Edition': 94,
 'Undertale': 92,
 'METAL GEAR SOLID V: THE PHANTOM PAIN': 91,
 'Pillars of Eternity': 89}